.buildkite/src/Command/Base.dhall

```diff
@@ -0,0 +1,52 @@
+-- This component hides the "useless" parts of the Buildkite command step bindings
+
+let Prelude = ../External/Prelude.dhall
+let Map = Prelude.Map
+let B = ../External/Buildkite.dhall
+in
+
+\(plugin : Type) ->
+
+let B/Command = B.definitions/commandStep/Type Text Text plugin plugin
+
+let Config = {
+  Type = {
+    agents : Optional (Map.Type Text Text),
+    artifact_paths : Optional (B.definitions/commandStep/properties/artifact_paths/Type),
+    branches : Optional (B.definitions/commandStep/properties/branches/Type),
+    commands : Optional (B.definitions/commandStep/properties/commands/Type),
+    depends_on : Optional (B.definitions/commandStep/properties/depends_on/Type),
+    `env` : Optional (Map.Type Text Text),
+    `if` : Optional Text,
+    key : Optional Text,
+    label : Optional Text,
+    plugins : Optional (B.definitions/commandStep/properties/plugins/Type plugin plugin),
+    retry : Optional (B.definitions/commandStep/properties/retry/Type),
+    timeout_in_minutes : Optional Natural
+  },
+  default = {
+    artifact_paths = None (B.definitions/commandStep/properties/artifact_paths/Type),
+    branches = None (B.definitions/commandStep/properties/branches/Type),
+    `env` = None (Map.Type Text Text),
+    `if` = None Text,
+    retry = None (B.definitions/commandStep/properties/retry/Type),
+    timeout_in_minutes = None Natural
+  }
+}
+
+let build : Config.Type -> B/Command = \(c : Config.Type) ->
+  c //
```

Comment:
These should all just be defaults for `B/Command`. As you are using this pattern right now, it seems like there is not need for a "Base" command as you are not injecting anything into it, you are just re-exposing the BuildKite config. Unless the goal here is to explicitly restrict access to these fields in our configuration, I'm not totally sure this makes sense. If your goal is the latter, there is a less verbose way to do this.

```
let Config = {
  Type = ...,
  default = ...
}

let build = \(c : Config.Type) -> B/Command::c
```

Reply:
Oops, I wish a read this more carefully during the day:

> These should all just be defaults for B/Command

Unfortunately, Buildkite's pipeline json-schema does not mark which fields are required, so I chose to take the more "safe" approach and assume everything is required from a generated binding perspective. I figured the reverse could cause us to miss fields by accident. I'd rather be forced to update this file when the schema changes than miss out on new fields. Happy to change this, if you have a good argument otherwise -- it will be an easy change in the schema compiler.

> the goal here is to explicitly restrict access to these fields in our configuration

Yes this is exactly the goal -- I wanted to add a layer of separation between the choice of which fields are required, optional, and skipped for our use-case. I will re-push and make that more clear. I felt that doing this alongside the other logic in what is now called `Command/Dsl.dhall` felt messy. I could also do this step in a helper function inside a since `Command/Base.dhall`, but I'm not sure that would be better.

> If your goal is the latter, there is a less verbose way to do this

I never considered passing a non-literal record after the `::` in a schema before, thanks for the tip. Unfortunately, it doesn't work in this case because of the lack of defaults. If we decide to add in the defaults above we can use that.

Reply:
In person, Nathan suggested we modify the buildkite schema to denote required things and generate Dhall schema files. Doing this, makes the explicit restriction bit not really worth this extra file.